demos/tests/cases.py

```diff
@@ -68,6 +71,60 @@ def fixed_args(self, source_dir, build_dir):
             *(['-l', str(cpu_extension_path)] if cpu_extension_path.exists() else [])]
 
 
+class NotebookDemo(Demo):
+    """
+    Converts notebook to a Python script with nbconvert. Adds code to make the
+    script work with command line parameters. The `test_code` parameter is
+    added at the end of the script. This should contain the line of code to
+    execute to test the notebook.
+    """
+
+    def __init__(self, name, implementation='jupyter_python', device_keys=None, test_cases=None, test_code=None):
+        super().__init__(name, implementation, device_keys, test_cases)
+        self._exec_name = self._exec_name.replace('_jupyter_python', '')
+        self.test_code = test_code
+
+    def fixed_args(self, source_dir, build_dir):
+        demo_env = {**os.environ,
+                   'PATH': f"{os.environ['PATH']}{os.pathsep}"
+                           f"{os.path.dirname(sys.executable)}",
+                   'PYTHONPATH': f"{os.environ['PYTHONPATH']}{os.pathsep}"
+                                 f"{os.path.join(os.path.dirname(os.path.abspath(os.curdir)), 'common', 'python')}"}
+
+        notebook_file = source_dir / self.subdirectory / (self._exec_name + '.ipynb')
+        python_file = notebook_file.with_suffix('.py')
+        python_test_file = str(python_file.with_suffix('')) + '_test.py'
+        subprocess.run([sys.executable, '-m', 'jupyter', 'nbconvert', '--to', 'python', str(notebook_file)], env=demo_env)
+
+        changedir_command = f"import os\nos.chdir(r'{os.path.dirname(python_file)}')\n"
+        original_content = python_file.read_text()
+        # Add `test_code` to the test script and make it accept command line arguments
+        content = inspect.cleandoc(f"""
+        # Change the working directory to the directory that contains the notebook
+        {changedir_command}
+        {original_content}
+
+        from os.path import dirname
+        import argparse
+
+        def parse_args():
+            parser = argparse.ArgumentParser()
+            parser.add_argument('-m')
+            parser.add_argument('-i')
+            return parser.parse_args()
+
+        args = parse_args()
+        # Change settings for test script
+        content += '\nbase_model_dir=dirname(dirname(dirname(dirname(args.m))))'
+        # Add a line that executes the function that is tested
+        content += f'\n{self.test_code}'
```

Comment:
You need to remove the assignments for this to work.

Reply:
Thanks, I noticed that too, I'll fix it today.

Reply:
@vladimir-dudnik @IRDonch I hope I fixed the tests in the CI. I kept the PATH and PYTHONPATH statements, even though I think they're unnecessary now, but I really want the CI tests to pass again first. Doublechecked on my local computer that run_tests.py works correctly.

Reply:
@IRDonch "Is it really required for the current directory to be the notebook directory for the imports to work?" (I cannot reply there) Yes. I can also solve it in some other way by adding a line to change the PYTHONPATH there, but I don't know if that's better.

Reply:
> I hope I fixed the tests in the CI. I kept the PATH and PYTHONPATH statements, even though I think they're unnecessary now, but I really want the CI tests to pass again first.

Well, CI did pass (and I don't think the failures were related to that change in the first place), so I suggest you remove them again.

> "Is it really required for the current directory to be the notebook directory for the imports to work?" (I cannot reply there) Yes.

But why? Now that I think about it, you're not even running a notebook; you're running a normal Python script. And in Python, the script directory is automatically added to the module search path. So why do you need to change the directory.

Reply:
OK, I see what's going on. You used `os.curdir` in the notebook to find the OMZ root. You should be using `__file__` instead - that way, you can run the notebook from any directory.

Reply:
That would solve this (and other issues) indeed, but `__file__`  is not defined in notebooks. I spend some time looking into this when I first made this notebook and there doesn't seem to be a solution that always works in the way it does for regular scripts (I tried quite a few things I found on Stack Overflow). Suggestions welcome.

Reply:
I see. That is quite annoying.

Well, in that case your solution is the best option you have. I would only request you to merge the `changedir` variable into the template, which would make it more readable.

Reply:
Thanks, I will do that, and remove the demo_env with the PATH variables. They were initially added with another change request, when the tests also would not run in the CI, but I think they are no longer necessary after the recent changes/new way to invoke Jupyter.